h/db/__init__.py

```diff
@@ -115,6 +115,22 @@ def close_the_sqlalchemy_session(request):
             })
         session.close()
 
+        # zope.sqlalchemy maintains an internal `id(session) => state` map with
+        # an entry for each active DB session which is registered with it.
+        #
+        # Entries are normally cleared at the end of a request when the
+        # transaction manager (`request.tm`) commits. DB writes after this can
+        # leave stale entries in the map which can cause problems in future
+        # requests if another session gets the same ID as the current one.
+        dm = zope.sqlalchemy.datamanager
+        if len(dm._SESSION_STATE) > 0:
```

Comment:
Out of interest, what’s the reason for this choice of `len(...) > 0`, rather than relying on Python’s native truthy/falsey evaluation?

Reply:
Oh, that's me thinking in JS where an empty dict/object (`{}`) is not falsey. Perhaps worth keeping this so that it is more likely to blow up in case `_SESSION_STATE` ever changes its type in future?

Reply:
> Perhaps worth keeping this so that it is more likely to blow up in case _SESSION_STATE ever changes its type in future?

Sure; I don’t see it doing any great harm, especially given it’s a private attribute in an external dependency.